File: python/algo/202512/algo0792.py
from bisect import bisect_right
from typing import List

# 九、子序列自动机
class Solution:
    def numMatchingSubseq(self, s: str, words: List[str]) -> int:
        # 参考题解
        pos = [[] for _ in range(26)]

        for i, ch in enumerate(s):
            pos[ord(ch)-ord('a')].append(i)
        
        print(pos)
        ans = 0
        for word in words:
            start = -1
            satisfied = True
            for ch in word:
                inx = bisect_right(pos[ord(ch)-ord('a')], start)
                if inx == len(pos[ord(ch)-ord('a')]):
                    satisfied = False
                    break
                else: start = pos[ord(ch)-ord('a')][inx]

            if satisfied: ans += 1
        
        return ans

File: cplus/leetcode/202512/algo0792.cpp
#include <bits/stdc++.h>
using namespace std;

// 九、子序列自动机
class Solution {
public:
    int numMatchingSubseq(string s, vector<string>& words) {
        // 参考题解
        vector pos(26, vector<int>());

        for (int i=0; i<s.length(); i++) pos[s[i]-'a'].emplace_back(i);

        int ans = 0;
        for (auto& word : words) {
            bool satisfied = true; int start = -1;
            for (auto & ch : word) {
                auto it = ranges::upper_bound(pos[ch-'a'], start);
                if (it != pos[ch-'a'].end()){
                    start = *it;
                } else {
                    satisfied = false; break;
                }
            }
            if (satisfied) ans ++;
        }

        return ans;     
    }
};